### PR TITLE
Add support for react-apollo@^3.0.0

### LINKED ISSUE #14: Export the OperationError type
Exporting https://github.com/molindo/react-apollo-network-status/blob/e08e7b43e2e3447ec0d9399262d17b162162805e/src/useApolloNetworkStatus.ts#L44 would be very useful for creating an error handling component. Currently I copy-pasted this type into my codebase which isn't ideal.

### LINKED ISSUE #16: Update to react-apollo 3.x
Installing this package results in this mismatch:

```
warning " > react-apollo-network-status@2.0.0" has incorrect peer dependency "react-apollo@^2.0.0".
```

`react-apollo` is currently at 3.1.1.

### PR DESCRIPTION
**Pull request by amannn**
Changes:
- We no longer depend on `react-apollo` to get the Apollo context (rather on `@apollo/react-common`). This makes it compatible with the new Apollo React integrations and allows for smaller bundle size in apps where you only use `@apollo/react-hooks`. Resolves #16 
- Export more types, resolves #14 
- Simplified examples. They're now using Hooks instead of render props.
- Upgraded dependencies.

Available as `react-apollo-network-status@3.0.0-alpha.1` on npm.

### CODE REVIEW

File: src/ApolloNetworkStatusProvider.tsx [hunk 1]
```diff
@@ -1,23 +1,24 @@
...
-  client,
-  children,
-  enableBubbling
-}: Props) {
+function ApolloNetworkStatusProvider({children, enableBubbling}: Props) {
+  const {client} = useContext(getApolloContext());
   const dispatcher = useMemo(() => new Dispatcher(), []);
 
+  if (!client) {
+    throw new Error(
```
Comment (amannn) [inline src/ApolloNetworkStatusProvider.tsx]: We'll probably never get to here when the context isn't defined. We should do this check before we destructure the `client`.

**Comment by amannn** [comment]
@msparer @symn Not sure if you also got a notification, but @Matth10 just opened a similar PR: https://github.com/molindo/react-apollo-network-status/pull/17.

I've asked him to help review this PR and he'll test the changes in his app.

**Comment by Matth10** [comment]
@amannn Seems that the version `3.0.0-alpha` works with my app know !

File: src/ApolloNetworkStatusProvider.tsx (hunk 1 above)
Comment (amannn) [reply src/ApolloNetworkStatusProvider.tsx]: Actually no. TypeScript is right, `getApolloContext` returns at least an empty object, but the `client` is optional.

File: package.json [hunk 2]
```diff
@@ -14,73 +14,71 @@
     "lint": "eslint './{src,example}/**/*.{ts,tsx}' && tsc --noEmit",
     "test": "jest",
     "build": "yarn lint && yarn test && rm -rf dist && yarn build:compile",
-    "build:compile": "microbundle --format cjs,umd",
+    "build:compile": "microbundle --format cjs,umd --globals @apollo/react-common=apolloReactCommon,apollo-link=apolloLink",
```
Comment (amannn) [inline package.json]: To be in sync with [the config from react-apollo](https://github.com/apollographql/react-apollo/blob/4a9f9b6bb15a2ec6dab95170757e00e2652f5d4b/config/rollup.config.js#L11-L33).

This allows to load the library unbundled as a script tag.

**Comment by rcohen-unext** [comment]
Just checking in to say 3.0.0-alpha.1 seems to be working great. The only thing is you are still using `useLayoutEffect` which is throwing a warning on SSR. I think you can safely change back to using `useEffect` with react-apollo 3.0+?

**Comment by amannn** [comment]
@rcohen-unext That's a good point, thanks for bringing this up!

File: src/useApolloNetworkStatusReducer.ts [hunk 3]
```diff
@@ -22,12 +22,11 @@ export default function useApolloNetworkStatusReducer<T>(
...
-  // Theoretically we should `useEffect` here. However react-apollo@<=2.5.5 uses
-  // React classes and queries can begin to fire in `componentDidMount`.
-  // Therefore we'd miss the request event of an operation. Once `react-apollo`
-  // has migrated, we could bump the peer dependency and use the correct hook
-  // here for a slight performance improvement.
-  useLayoutEffect(() => {
+  // Effects fire bottom-up. Therefore it's possible that when a query is nested
+  // further down the tree than a component using this hook and the query fires
+  // on the initial render, we'll miss the request event. Note that this isn't
+  // an issue when pre-rendering or -fetching on the server side.
```
Comment (amannn) [inline src/useApolloNetworkStatusReducer.ts]: There's theoretically a fix for this, but it increases complexity a bit. Seems like this has never been an issue so far and if it comes up, we can still fix it in a patch release.

Comment (amannn) [reply src/useApolloNetworkStatusReducer.ts]: Actually here's the request for this: https://github.com/molindo/react-apollo-network-status/issues/19#issue-513171800. It shouldn't block this PR IMO though, as the situation is the same before and after this PR.

**Comment by MasterKale** [comment]
Are there any known compatibility issues with this PR and the `Provider` from **@apollo/react-hooks** or `ApolloClient` from **apollo-boost**? I tried installing **react-apollo-network-status@3.0.0alpha-1** into an existing project using both of these other dependencies but for some reason my "NetworkIndicator" Function Component using the `useApolloNetworkStatus()` hook never seems to get updated network status.

Here's a Codesandbox demonstrating this. If you pop open the in-editor Console, you'll see that the `console.log()` in `NetworkIndicator` only gets called once:

https://codesandbox.io/s/jovial-fire-kgcmr?fontsize=14

Am I doing something wrong here?

**Comment by amannn** [comment]
@MasterKale Thanks for the bug report and the CodeSandbox reproduction.

I had a look and as you've mentioned, I think the issue is that they're [extending the regular Apollo Client](https://github.com/apollographql/apollo-client/blob/5082138ae82f7d08dc69e2ba388854438ed31b43/packages/apollo-boost/src/index.ts#L75) and that breaks some of the functionality that we're using in this library.

I think for [the upcoming 3.0 version of Apollo Client](https://github.com/apollographql/apollo-client/blob/master/ROADMAP.md#goals), `apollo-boost` will be deprecated since the main `apollo-client` package will already contain all the necessary dependencies to get started. So I think I wouldn't spend effort on that compatibility for now. I think already the 2.0 version of this library doesn't work with `apollo-boost` for the same reason.

Do you have the option to use the regular `apollo-client` in your project? Then the upcoming version of this library should work for you.

**Comment by amannn** [comment]
@msparer Do you have any feedback for this PR or is it ok for you if I merge?

**Pull request merged**